R/02-getData.R

```diff
@@ -0,0 +1,250 @@
+#' Get Data
+#'
+#' @param name (character) name of a resource, e.g. an entry of the output from
+#'  \code{getResources()$name}
+#' @param options (list) a list of extra options for \code{read.csv()} or \code{openxlsx::read.xlsx()} and
+#'  \code{readxl::read_excel}
+#' @inheritParams getResources
+#'
+#' @return (data.frame) return data from the Pandora API
+#' @export
+getData <- function(name,
+                    repository = "",
+                    options = dataOptions()) {
+  # empty result if an error occurs
+  res <- list()
+  
+  # filter repository
+  resourcesForRepo <- getResources(repository = repository)
+  if (nrow(resourcesForRepo) == 0) {
+    attr(res, "error") <-
+      sprintf("No resource found for repository '%s'", repository)
+    return(res)
+  }
+  
+  # filter name
+  resource <- resourcesForRepo[resourcesForRepo[["name"]] == name, ]
+  if (nrow(resource) == 0) {
+    attr(res, "error") <-
+      sprintf("No resource found with name '%s'", name)
+    return(res)
+  }
+  
+  # filter valid file types
+  resource <-
+    resource[resource[["format"]] %in% config()$fileTypes, ]
+  if (nrow(resource) == 0) {
+    attr(res, "error") <-
+      sprintf(
+        "No resource found with name '%s' and with valid file type (%s)",
+        name,
+        paste(config()$fileTypes, collapse = ", ")
+      )
+    return(res)
+  }
+  
+  # select single file
+  if (nrow(resource) > 1) {
+    # if more than one file for one name, order by config()$fileTypes
+    orderVec <-
+      na.omit(match(config()$fileTypes, resource[["format"]]))
+    resource <- resource[orderVec, ]
+    # take only first file
+    resource <- resource[1, ]
+  }
+  
+  data <- try({
+    loadData(
+      path = resource[["url"]],
+      type = resource[["format"]],
+      nrows = options$nrows,
+      sep = options$text$sep,
+      dec = options$text$dec,
+      colNames = options$colNames,
+      sheet = options$xlsx$sheet
+    )
+  }, silent = TRUE)
+  
+  if (inherits(data, "try-error")) {
+    msg <- ""
+    if (resource[["format"]] == "csv")
+      msg <- "Please check dataOptions()."
+    msg <-
+      sprintf("%s for resource with name '%s', %s", data[[1]], name, msg)
+    warning(msg)
+    attr(res, "error") <- msg
+  } else if (!is.null(data) && length(data) > 0 && nrow(data) > 0) {
+    # data loading SUCCESS
+    res <- data
+  } else if (length(data) == 0 && !is.null(attr(data, "error"))) {
+    msg <-
+      sprintf("%s for resource with name '%s'", attr(data, "error"), name)
+    attr(res, "error") <- msg
+  } else {
+    msg <- sprintf("An error occured for resource with name '%s'", name)
+    warning(msg)
+    attr(res, "error") <- msg
+  }
+  
+  res
```

Comment:
Now the `getData` function should be way more readable

```suggestion
 #' Get Data
#'
#' @param name (character) name of a resource, e.g. an entry of the output from
#'  \code{getResources()$name}
#' @param options (list) a list of extra options for \code{read.csv()} or \code{openxlsx::read.xlsx()} and
#'  \code{readxl::read_excel}
#' @inheritParams getResources
#'
#' @return (data.frame) return data from the Pandora API
#' @export
getData <- function(name,
                    repository = "",
                    options = dataOptions()) {
                    
  resourcesForRepo <- getResources(repository = repository)
  
  filterResourceByRepo(...) # i did not give the `first function` a name - (first review comment)
  filterResourceByName(name, resourcesForRepo)
 
  resource <- filterValidFileType(resourcesForRepo)
  resource <- selectSingleFile(resource)
  
  loadDataFromResource(resource, options)
}
```